### PR TITLE
Enable testing on Windows with AppVeyor

### PR DESCRIPTION
**Pull request by wilzbach**
With the `build.bat`

### CODE REVIEW

File: build.bat [hunk 1]
```diff
@@ -1,30 +1,31 @@
...
-set STD=
-set ANALYSIS=
-set INIFILED=
-set DSYMBOL=
-set CONTAINERS=
-set LIBDDOC=
-
-for %%x in (src\*.d) do set CORE=!CORE! %%x
-for %%x in (src\analysis\*.d) do set ANALYSIS=!ANALYSIS! %%x
-for %%x in (libdparse\experimental_allocator\src\std\experimental\allocator\*.d) do set STD=!STD! %%x
```
Comment (BBasile) [inline build.bat]: line to remove, also next line

File: build.bat [hunk 2]
```diff
@@ -1,30 +1,31 @@
...
-setlocal enabledelayedexpansion
-
-set DFLAGS=-O -release -inline
-set CORE=
-set LIBDPARSE=
-set STD=
-set ANALYSIS=
-set INIFILED=
-set DSYMBOL=
-set CONTAINERS=
```
Comment (BBasile) [inline build.bat]: to remove

File: build.bat [hunk 3]
```diff
@@ -1,30 +1,31 @@
...
-for %%x in (libddoc\src\ddoc\*.d) do set LIBDDOC=!LIBDDOC! %%x
-for %%x in (inifiled\source\*.d) do set INIFILED=!INIFILED! %%x
-for %%x in (dsymbol\src\dsymbol\*.d) do set DSYMBOL=!DSYMBOL! %%x
-for %%x in (dsymbol\src\dsymbol\builtin\*.d) do set DSYMBOL=!DSYMBOL! %%x
-for %%x in (dsymbol\src\dsymbol\conversion\*.d) do set DSYMBOL=!DSYMBOL! %%x
-for %%x in (containers\src\containers\*.d) do set CONTAINERS=!CONTAINERS! %%x
-for %%x in (containers\src\containers\internal\*.d) do set CONTAINERS=!CONTAINERS! %%x
-
-@echo on
-dmd %CORE% %STD% %LIBDPARSE% %LIBDDOC% %ANALYSIS% %INIFILED% %DSYMBOL% %CONTAINERS% %DFLAGS% -I"libdparse\src" -I"dsymbol\src" -I"containers\src"  -I"libddoc\src" -ofdscanner.exe
```
Comment (BBasile) [inline build.bat]: and `%CONTAINERS%` too.

File: build.bat [hunk 4]
```diff
@@ -1,30 +1,31 @@
...
-
+@echo off
+setlocal enabledelayedexpansion
+
+IF "%DC%"=="" SET DC="dmd"
+set DFLAGS=-O -release -inline
+set CORE=
+set LIBDPARSE=
+set STD=
```
Comment (BBasile) [inline build.bat]: remove

File: build.bat [hunk 5]
```diff
@@ -1,30 +1,31 @@
...
+set ANALYSIS=
+set INIFILED=
+set DSYMBOL=
+set CONTAINERS=
+set LIBDDOC=
+
+for %%x in (src\*.d) do set CORE=!CORE! %%x
+for %%x in (src\analysis\*.d) do set ANALYSIS=!ANALYSIS! %%x
+for %%x in (libdparse\experimental_allocator\src\std\experimental\allocator\*.d) do set STD=!STD! %%x
```
Comment (BBasile) [inline build.bat]: remove

File: build.bat [hunk 6]
```diff
@@ -1,30 +1,31 @@
...
+for %%x in (libdparse\experimental_allocator\src\std\experimental\allocator\building_blocks\*.d) do set STD=!STD! %%x
```
Comment (BBasile) [inline build.bat]: remove

File: build.bat [hunk 7]
```diff
@@ -1,30 +1,31 @@
...
+for %%x in (libddoc\src\ddoc\*.d) do set LIBDDOC=!LIBDDOC! %%x
+for %%x in (inifiled\source\*.d) do set INIFILED=!INIFILED! %%x
+for %%x in (dsymbol\src\dsymbol\*.d) do set DSYMBOL=!DSYMBOL! %%x
+for %%x in (dsymbol\src\dsymbol\builtin\*.d) do set DSYMBOL=!DSYMBOL! %%x
+for %%x in (dsymbol\src\dsymbol\conversion\*.d) do set DSYMBOL=!DSYMBOL! %%x
+for %%x in (containers\src\containers\*.d) do set CONTAINERS=!CONTAINERS! %%x
+for %%x in (containers\src\containers\internal\*.d) do set CONTAINERS=!CONTAINERS! %%x
+
+@echo on
+%DC% %CORE% %STD% %LIBDPARSE% %LIBDDOC% %ANALYSIS% %INIFILED% %DSYMBOL% %CONTAINERS% %DFLAGS% -I"libdparse\src" -I"dsymbol\src" -I"containers\src"  -I"libddoc\src" -ofdscanner.exe
```
Comment (BBasile) [inline build.bat]: remove `%STD%`

File: build.bat [hunk 8]
```diff
@@ -1,30 +1,31 @@
...
-for %%x in (containers\src\containers\internal\*.d) do set CONTAINERS=!CONTAINERS! %%x
-
-@echo on
-dmd %CORE% %STD% %LIBDPARSE% %LIBDDOC% %ANALYSIS% %INIFILED% %DSYMBOL% %CONTAINERS% %DFLAGS% -I"libdparse\src" -I"dsymbol\src" -I"containers\src"  -I"libddoc\src" -ofdscanner.exe
-
+@echo off
+setlocal enabledelayedexpansion
+
+IF "%DC%"=="" SET DC="dmd"
+set DFLAGS=-O -release -inline
```
Comment (BBasile) [inline build.bat]: DFlags should change according to DC. For ldc they should be set to `-O2 -release` (O2 imply inlining)

Comment (BBasile) [inline build.bat]: another option is to remove `-inline`. Your call...

**Comment by wilzbach** [comment]
> DFlags should change according to DC. For ldc2 they should be set to -O2 -release (O2 set implicitly inlining on)
> another option is to remove -inline. Your call...

I used another trick:

```
IF "%DC%"=="ldc2" SET DC="ldmd2"
```

In theory this should also fix the issues with the invalid `-lib` flag ...

File: appveyor.yml [hunk 9]
```diff
@@ -0,0 +1,123 @@
+platform: x64
+environment:
+ matrix:
+  - DC: dmd
+    DVersion: nightly
```
Comment (BBasile) [inline appveyor.yml]: ~~nightly~~

File: appveyor.yml [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    arch: x64
+  - DC: dmd
+    DVersion: nightly
```
Comment (BBasile) [inline appveyor.yml]: ~~nightly~~

**Comment by BBasile** [comment]
You can now run the tests under windows with `build.bat test`.

**Comment by wilzbach** [comment]
So now it seems that we run into a lot of funky messages: https://ci.appveyor.com/project/greenify/d-scanner/build/1.0.16/job/34hu2tnnwa23yw4y, e.g. 

```
core.exception.AssertError@src\analysis\alias_syntax_check.d(48): Expected warning:
[warn]: Prefer the new "'alias' identifier '=' type ';'" syntax to the  old "'alias' type identifier ';'" syntax.
		alias abcde = int;
	
From source code at (48:?):
		alias int abcde; // [warn]: Prefer the new "'alias' identifier '=' type ';'" syntax to the  old "'alias' type identifier ';'" syntax.
		alias abcde = int;
	
----------------
0x00007FF6826BD41F in analysis.helpers.assertAnalyzerWarnings.__foreachbody5 at C:\projects\d-scanner\src\analysis\helpers.d(112)
```

As this hasn't been there this summer, I would assume this is due to a recent change in the compiler. 
Is anyone using D-Scanner on Windows and wants to look into this?
(I don't have a Windows machine)

File: build.bat [hunk 11]
```diff
@@ -1,44 +1,47 @@
...
-
-if exist %TESTNAME%.obj del %TESTNAME%.obj
-if exist %TESTNAME%.lib del %TESTNAME%.lib
-if exist %TESTNAME%.exe del %TESTNAME%.exe
-
-:eof
+@echo off
+setlocal enabledelayedexpansion
+
+IF "%DC%"=="" SET DC="dmd"
```
Comment (BBasile) [inline build.bat]: Sorry but can you write the command in lower case ?
Also do you know why there's a major diff failure for this file ?

**Comment by BBasile** [comment]
> As this hasn't been there this summer, I would assume this is due to a recent change in the compiler.
Is anyone using D-Scanner on Windows and wants to look into this?
(I don't have a Windows machine)

No, the last change in this area [is two years old](https://github.com/dlang-community/D-Scanner/blame/master/src/analysis/alias_syntax_check.d#L49), the problem is more likely that the config file has changed. Here unittests should be skipped since we explicitly use the old syntax to verify that a warning is emitted.

Actually a logic rule would be to always set skip_unittest in the filters and when a check check its own implementation.

**Comment by BBasile** [comment]
(previous comment deleted because) Wait...this is a string actually.

**Comment by BBasile** [comment]
I'm on win right now. `build.bat test` passed with 2.076.1 and 2.077.1

**Comment by wilzbach** [comment]
> I'm on win right now. build.bat test passed with 2.076.1 and 2.077.1

Thanks a lot for the info... This is great :/
Do you have other ideas?

> Also do you know why there's a major diff failure for this file ?

I assume because my Vim automatically fixed the line-endings?
It's still weird though.

**Comment by BBasile** [comment]
I'll try to setup Appveyor on my fork and see. D-Scanner doesn't fundamentally requires multi platform tests but this is so weird that an explanation must be found.

**Comment by skl131313** [comment]
@wilzbach 
Seems LDC changed their naming convention for releases, your download script doesn't work. Also seems build.bat need to be updated to support x64.

**Comment by wilzbach** [comment]
> Seems LDC changed their naming convention for releases, your download script for the ldc compiler is broken cause of that.

Thanks a lot for getting back to this & finding the root cause. I simply disabled `ldc-beta` and LDC x86 builds for now.
It will be a lot easier to re-enable them on a separate PR, testing with DMD on Windows already brings a great value and avoids the annoying red cross.

**Pull request merged**